File: mixql-protobuf/src/main/scala/org/mixql/protobuf/GtypeConverter.scala
package org.mixql.protobuf

object GtypeConverter {

  import org.mixql.core.context.gtype
  import org.mixql.protobuf.generated.messages

  def toGtype(remoteMsg: com.google.protobuf.GeneratedMessageV3): gtype.Type = {
    import collection.JavaConverters._

    remoteMsg match {
      case _: messages.NULL => gtype.Null
      case msg: messages.Bool => gtype.bool(msg.getValue)
      case msg: messages.Int => gtype.int(msg.getValue)
      case msg: messages.Double => gtype.double(msg.getValue)
      case msg: messages.String => gtype.string(msg.getValue)
      case msg: messages.Array =>
        gtype.array(
          msg.getArrList.asScala
            .map(f => protobufAnyToGtype(f))
            .toArray
        )
      case msg: messages.Error => throw new Exception(msg.getMsg)
      case a: scala.Any =>
        throw new Exception(
          s"RemoteMsgsConverter: toGtype error: " +
            s"got ${a.toString}, when type was expected"
        )
    }
  }

  def protobufAnyToGtype(f: com.google.protobuf.Any): gtype.Type = {
    if (f.is(messages.NULL.getDefaultInstance.getClass))
      return gtype.Null

    if (f.is(messages.Bool.getDefaultInstance.getClass)) {
      val msg = f.unpack(messages.Bool.getDefaultInstance.getClass)
      return gtype.bool(msg.getValue)
    }

    if (f.is(messages.Int.getDefaultInstance.getClass)) {
      val msg = f.unpack(messages.Int.getDefaultInstance.getClass)
      return gtype.int(msg.getValue)
    }

    if (f.is(messages.Double.getDefaultInstance.getClass)) {
      val msg = f.unpack(messages.Double.getDefaultInstance.getClass)
      return gtype.double(msg.getValue)
    }

    if (f.is(messages.String.getDefaultInstance.getClass)) {
      val msg = f.unpack(messages.String.getDefaultInstance.getClass)
      return gtype.string(msg.getValue)
    }


    if (f.is(messages.Array.getDefaultInstance.getClass)) {
      val msg = f.unpack(messages.Array.getDefaultInstance.getClass)
      return toGtype(msg)
    }

    throw new Exception(
      s"protobufAnyToGtype: error:  " +
        s"could not convert com.google.protobuf.any.Any to gtype: got ${f.toString}, when gtype was expected"
    )
  }

  def toGeneratedMsg(gValue: gtype.Type): com.google.protobuf.GeneratedMessageV3 = {
    gValue match {
      case gtype.Null =>
        messages.NULL.newBuilder().build()
      case gtype.bool(value) =>
        messages.Bool.newBuilder().setValue(value).build()
      case gtype.int(value) =>
        messages.Int.newBuilder().setValue(value).build()
      case gtype.double(value) =>
        messages.Double.newBuilder().setValue(value).build()
      case gtype.string(value, quote) =>
        messages.String.newBuilder()
          .setValue(value)
          .setQuote(quote)
          .build()
      case gtype.array(arr) =>
        var arrayBuilder = messages.Array.newBuilder

        arr.map(gType =>
          com.google.protobuf.Any.pack(toGeneratedMsg(gType))
        ) foreach (
          v => arrayBuilder = arrayBuilder.addArr(v)
          )
        arrayBuilder.build()
    }
  }

  def toProtobufAny(remoteMsg: com.google.protobuf.GeneratedMessageV3): com.google.protobuf.Any = {
    remoteMsg match {
      case _: messages.NULL => com.google.protobuf.Any.pack(messages.NULL.getDefaultInstance)
      case msg: messages.Bool => com.google.protobuf.Any.pack(msg)
      case msg: messages.Int => com.google.protobuf.Any.pack(msg)
      case msg: messages.Double => com.google.protobuf.Any.pack(msg)
      case msg: messages.String => com.google.protobuf.Any.pack(msg)
      case msg: messages.Array => com.google.protobuf.Any.pack(msg)
      case msg: messages.Error => throw new Exception(msg.getMsg)
      case a: scala.Any =>
        throw new Exception(
          s"RemoteMsgsConverter: toProtobufAny error: " +
            s"got ${a.toString}, when type in scalapb.GeneratedMessage format was expected"
        )
    }
  }

  def toProtobufAny(gValue: gtype.Type): com.google.protobuf.Any = {
    gValue match {
      case gtype.Null =>
        com.google.protobuf.Any.pack(messages.NULL.newBuilder().build())
      case gtype.bool(value) =>
        com.google.protobuf.Any.pack(messages.Bool.newBuilder().setValue(value).build())
      case gtype.int(value) =>
        com.google.protobuf.Any.pack(messages.Int.newBuilder().setValue(value).build())
      case gtype.double(value) =>
        com.google.protobuf.Any.pack(messages.Double.newBuilder().setValue(value).build())
      case gtype.string(value, quote) =>
        com.google.protobuf.Any.pack(messages.String.newBuilder()
          .setValue(value).setQuote(quote).build()
        )
      case gtype.array(arr) =>
        var arrayBuilder = messages.Array.newBuilder

        arr.map(gType =>
          com.google.protobuf.Any.pack(toGeneratedMsg(gType))
        ) foreach (
          v => arrayBuilder = arrayBuilder.addArr(v)
          )

        com.google.protobuf.Any.pack(
          arrayBuilder.build()
        )
    }
  }
}

File: mixql-protobuf/src/main/java/org/mixql/protobuf/GtypeConverter.java
package org.mixql.protobuf;

import com.google.protobuf.Any;
import org.mixql.core.context.gtype.*;
import org.mixql.protobuf.generated.messages.*;
import org.mixql.protobuf.generated.messages.Double;
import org.mixql.protobuf.generated.messages.Error;
import org.mixql.protobuf.generated.messages.String;

import java.util.*;

public class GtypeConverter {

    public static Type toGtype(com.google.protobuf.GeneratedMessageV3 remoteMsg) throws Exception {
        if (remoteMsg instanceof NULL)
            return new Null();

        if (remoteMsg instanceof Int)
            return new gInt(((Int) remoteMsg).getValue());

        if (remoteMsg instanceof Double)
            return new gDouble(((Double) remoteMsg).getValue());

        if (remoteMsg instanceof Bool)
            return new bool(((Bool) remoteMsg).getValue());

        if (remoteMsg instanceof String) {
            String msg = (String) remoteMsg;
            return new string(msg.getValue(), msg.getQuote());
        }

        if (remoteMsg instanceof Array) {
            List<Any> anyList = ((Array) remoteMsg).getArrList();
            Type[] gtypeList = new Type[anyList.size()];

            int i = 0;
            for (Any anyValue : anyList) {
                gtypeList[i++] = protobufAnyToGtype(anyValue);
            }

            return new array(gtypeList);
        }

        if (remoteMsg instanceof Error)
            throw new Exception(((Error) remoteMsg).getMsg());

        throw new Exception(
                java.lang.String.format(
                        "RemoteMsgsConverter: toGtype error: " +
                                "got unknown message %s", remoteMsg.getClass().getName()
                )
        );
    }

    public static Type protobufAnyToGtype(com.google.protobuf.Any f) throws Exception {
        System.out.println("mixql-protobuf: protobufAnyToGtype");
        if (f.is(NULL.getDefaultInstance().getClass()))
            return new Null();

        if (f.is(Bool.getDefaultInstance().getClass())) {
            Bool msg = f.unpack(Bool.getDefaultInstance().getClass());
            return new bool(msg.getValue());
        }

        if (f.is(Int.getDefaultInstance().getClass())) {
            Int msg = f.unpack(Int.getDefaultInstance().getClass());
            return new gInt(msg.getValue());
        }

        if (f.is(Double.getDefaultInstance().getClass())) {
            Double msg = f.unpack(Double.getDefaultInstance().getClass());
            return new gDouble(msg.getValue());
        }

        if (f.is(String.getDefaultInstance().getClass())) {
            String msg = f.unpack(String.getDefaultInstance().getClass());
            return new string(msg.getValue(), msg.getQuote());
        }


        if (f.is(Array.getDefaultInstance().getClass())) {
            Array msg = f.unpack(Array.getDefaultInstance().getClass());
            return toGtype(msg);
        }

        throw new Exception(
                java.lang.String.format(
                        "protobufAnyToGtype: error: could not convert " +
                                "com.google.protobuf.any.Any to gtype: got %s, when gtype was expected",
                        f.getClass().getName()
                )
        );
    }

    public static com.google.protobuf.GeneratedMessageV3 toGeneratedMsg(Type gValue) throws Exception {
        System.out.println("mixql-protobuf: toGeneratedMsg");
        if (gValue instanceof Null)
            return NULL.newBuilder().getDefaultInstanceForType();

        if (gValue instanceof bool)
            return Bool.newBuilder().setValue(((bool) gValue).getValue()).build();

        if (gValue instanceof gInt)
            return Int.newBuilder().setValue(((gInt) gValue).getValue()).build();

        if (gValue instanceof gDouble)
            return Double.newBuilder().setValue(((gDouble) gValue).getValue()).build();

        if (gValue instanceof string) {
            string msg = (string) gValue;
            return String.newBuilder().setValue(msg.getValue()).setQuote(msg.getQuote()).build();
        }

        if (gValue instanceof array) {
            Type[] gtypeArr = ((array) gValue).getArr();
            List<Any> anyValueList = new ArrayList<>();

            for (Type gType : gtypeArr) {
                anyValueList.add(Any.pack(toGeneratedMsg(gType)));
            }
            return Array.newBuilder().addAllArr(anyValueList).build();
        }

        throw new Exception(
                java.lang.String.format(
                        "toGeneratedMsg: error: could not convert " +
                                "gtype to com.google.protobuf.any.Any: got unknown %s",
                        gValue.getClass().getName()
                )
        );
    }

    public static Any toProtobufAny(com.google.protobuf.GeneratedMessageV3 remoteMsg) throws Exception {
        if (remoteMsg instanceof NULL)
            return Any.pack(NULL.getDefaultInstance());

        if (remoteMsg instanceof Bool)
            return Any.pack((Bool) remoteMsg);

        if (remoteMsg instanceof Int)
            return Any.pack((Int) remoteMsg);

        if (remoteMsg instanceof Double)
            return Any.pack((Double) remoteMsg);

        if (remoteMsg instanceof String)
            return Any.pack((String) remoteMsg);

        if (remoteMsg instanceof Array)
            return Any.pack((Array) remoteMsg);

        if (remoteMsg instanceof Error)
            throw new Exception(((Error) remoteMsg).getMsg());


        throw new Exception(
                java.lang.String.format(
                        "RemoteMsgsConverter: toProtobufAny error: " +
                                "got %s, when type in GeneratedMessageV3 format was expected",
                        remoteMsg.getClass().getName()
                )
        );
    }

    public static com.google.protobuf.Any toProtobufAny(Type gValue) throws Exception {
        if (gValue instanceof Null)
            return Any.pack(NULL.getDefaultInstance());

        if (gValue instanceof bool)
            return Any.pack(Bool.newBuilder().setValue(((bool) gValue).getValue()).build());

        if (gValue instanceof gInt)
            return Any.pack(Int.newBuilder().setValue(((gInt) gValue).getValue()).build());

        if (gValue instanceof gDouble)
            return Any.pack(Double.newBuilder().setValue(((gDouble) gValue).getValue()).build());

        if (gValue instanceof string) {
            string msg = (string) gValue;
            return com.google.protobuf.Any.pack(String.newBuilder()
                    .setValue(msg.getValue()).setQuote(msg.getQuote()).build()
            );
        }

        if (gValue instanceof array) {
            Type[] gtypeArr = ((array) gValue).getArr();
            List<Any> anyValueList = new ArrayList<>();

            for (Type gType : gtypeArr) {
                anyValueList.add(Any.pack(toGeneratedMsg(gType)));
            }

            return Any.pack(Array.newBuilder().addAllArr(anyValueList).build());
        }

        throw new Exception(
                java.lang.String.format(
                        "RemoteMsgsConverter: toProtobufAny error: " +
                                "got %s, when type in gValue format was expected",
                        gValue.getClass().getName()
                )
        );

    }
}
